### PR TITLE
Add a generic abstraction for filesystem based stores

### LINKED ISSUE #3365: Containers across namespaces boundary show up in /etc/hosts
### Description

```
nerdctl run -d --namespace one --name deb1 debian sleep Inf
nerdctl run -d --namespace two --name deb2 debian sleep Inf
nerdctl exec --namespace two deb2 cat /etc/hosts
```

This can lead to situations where entries in /etc/hosts are duplicated (when containers in different namspaces have the same name).

Of course this also breaks the promise of some isolation between ns.

### Steps to reproduce the issue

na

### Describe the results you received and expected

na

### What version of nerdctl are you using?

main

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

_No response_

### PR DESCRIPTION
**Pull request by apostasie**
Although our codebase makes use of flock when we manipulate the filesystem (normally...) , we have had a relatively significant number of issues in the past with our various "file stores". They typically fall in the following categories:
- unprotected concurrent access to the same resource
- filesystem details leaking upward into higher level abstractions where they are manipulated unsafely
- lock that does not really lock
- lock that should lock over a larger section of code
- sequential (individually locked) operations being done assuming that data cannot change in-between them
- partial writes to the filesystem

Issues arising from these are generally heinous to debug, as they usually manifest in a flaky way (as they may only happen in high-concurrency situations / under pressure), or tend to bubble-up in different places making seemingly unrelated operations fail in cryptic ways.

Truth is, it is not 100% trivial to achieve a concurrency-safe and atomic filesystem storage system, even a very simple one like what we do - operating system crash or restarts, or generally interrupted writes, for example, are often overlooked as something that has to be thought about, to boot.

While we have fixed a number of these issues, there is clearly more lingering, and we are bound to produce even more in the future short of having a stronger basis.
I do not think it is fair to assume that everyone will be very mindful and intentional with their locking, or that the filesystem is a good enough, "safe" abstraction...

Furthermore, and even when done right, these things are inherently repetitive (mutex / filesystem lock / path sanitization / atomic write) and we now have enough of these "filestores" to justify having a shared base:
- namestore
- volumestore
- lifecycledata
- hostsstore

**This PR first commit does introduce an interface and a reference (filesystem) implementation** of a reusable `Store` component that every "store" can leverage and that provides:
- concurrency safety (thread safe and safe across concurrent invocations of the binary)
- atomic operations (write will never leave incomplete data in a file)
- path components validation, per-platform
- a small set of simple methods that should cover most use-cases
- completely sealed underlying implementation that does not leak
- proper, detailed errors
- some level of locking enforcement (eg: will error if there is no lock)
- tests

Structs that leverage it are thus freed from having to deal with mutex / filelocks, or underlying implementations (eg: filesystem). They can rely on a simple API that is responsible for just that, and can focus solely on the abstraction they have to provide for their specific data.

Hopefully, if things go our way, we should never see again a bug report about hosed json files / broken systems on restart, or flakyness involving stores...

**The second commit is a mere cleanup of the (useless) mock for the volume store**

**The third commit is a refactor of our various stores** to leverage that:
- all four stores have been reviewed and their syntax harmonized
- better erroring
- code duplication removal
- all explicit locking mechanisms have been removed (to simplify the user life) (with the exception of the VolumeStore) and all stores now expose methods tailored for their use cases that lock on their own
- all use cases have been reviewed to make sure nothing is racy (exceptions with Composer, obviously)
- ensuring locking is done at the namespace level, or lower

Incidentally, the following known issues are getting fixed or partially fixed:
- fix #3365
- part of #3325

Finally, I do appreciate this is a far reaching PR, as it is touching code in the critical path of basically everything.
To a large extent, though, this is mere refactoring and literally no novelty.

### CODE REVIEW

File: cmd/nerdctl/namespace.go
Comment (apostasie) [inline cmd/nerdctl/namespace.go:122]: List() is expensive.
Replaced with the much lighter Count.

File: pkg/cmd/compose/compose.go
Comment (apostasie) [inline pkg/cmd/compose/compose.go:81]: There is now a `Exists` method, much cheaper than `Get`

File: pkg/cmd/container/remove.go
Comment (apostasie) [inline pkg/cmd/container/remove.go:135]: Removing the lock in that case (see comment).

File: pkg/cmd/volume/prune.go
Comment (apostasie) [inline pkg/cmd/volume/prune.go:36]: Replacing all of that with a `volStore.Prune(func)` that does locking.

File: pkg/identifiers/validate_windows.go
Comment (apostasie) [inline pkg/identifiers/validate_windows.go:1]: Moved inside the store.
These things are restrictions linked to the filesystem implementation, and not conceptually to restrictions we necessarily want to apply to identifiers.
Put otherwise: they are orthogonal concerns.

File: pkg/mountutil/mountutil_test.go
Comment (apostasie) [inline pkg/mountutil/mountutil_test.go:1]: This replace the whole uber go mock thing for mountutil.

**Comment by apostasie** [comment]
This is ready for review and CI is green (save the Kube flaky test).
This will need to be rebased atop of #3362 that presumably should get merged first.

File: pkg/store/filestore.go
Comment (apostasie) [inline pkg/store/filestore.go:240]: Will remove before merging.
This is not used.

File: pkg/store/store.go
Comment (apostasie) [inline pkg/store/store.go:88]: Will remove before merge. Unused.

File: pkg/cmd/container/remove.go
Comment (apostasie) [inline pkg/cmd/container/remove.go:132]: Calling volume.Store is useless in that case, as we already have dataStore.

File: pkg/cmd/container/rename.go
Comment (apostasie) [inline pkg/cmd/container/rename.go:47]: HostsStore signature got changed to be per-namespace - align with other stores, and guarantees better isolation between namespaces and less "global" locking.

File: pkg/cmd/container/run_mount.go
Comment (apostasie) [inline pkg/cmd/container/run_mount.go:261]: The Create method now provides locking.
CreateWithoutLock emphasizes that this method MUST be guardrailed inside a locked section.

File: pkg/containerutil/container_network_manager.go
Comment (apostasie) [inline pkg/containerutil/container_network_manager.go:404]: This property is useless (and was before IIRC).
The meta is being saved under a namespaced location in the store already.

File: pkg/containerutil/container_network_manager.go
Comment (apostasie) [inline pkg/containerutil/container_network_manager.go:513]: Write only once, instead of allocating first, then writing.

File: pkg/identifiers/validate.go
Comment (apostasie) [inline pkg/identifiers/validate.go:45]: Platform validation of identifiers is a filesystem concern.
These are moved into the `store.Store`, as they are orthogonal concerns.

File: cmd/nerdctl/volume_remove_linux_test.go
Comment (AkihiroSuda) [inline cmd/nerdctl/volume_remove_linux_test.go:44]: ?

File: pkg/cmd/volume/rm.go
Comment (apostasie) [inline pkg/cmd/volume/rm.go:35]: Thinking more about this, the entire section starting with client.Containers should be inside the locked segment.
Will amend.

File: cmd/nerdctl/volume_remove_linux_test.go
Comment (apostasie) [reply cmd/nerdctl/volume_remove_linux_test.go:44]: Yeah, been back and forth on this.
I still think we have a general problem in the codebase with regard to errors management.
Since we generally do `fmt.Errorf` instead of using exported errors (`var ErrFoo = errors.New`), consuming code cannot test with `errors.Is`, making it hard to specialize behavior dependent on the type of error we get.

Some of my recent PR _do_ introduce better, finer grain and explicit error-ing.
Obviously, some of the errors should be shared across different sections of code, maybe into an `errs` subpackage.

Which brings the question of whether we want to try and reuse some (or all) of `containerd/errdefs` (for example `ErrNotFound`).

I have mixed feelings about whether we should do that or not, and I think I am leaning towards having our own independent `errs` package, as it will make things clearer that what nerdctl returns is nerdctl errors (even if down there, they happen to alias containerd errdef).

Thoughts?

**Comment by apostasie** [comment]
Rebase broke some tests.
Investigating.

**Comment by apostasie** [comment]
- rebased
- removed commented out extra stuff
- changed `volStore.Remove` semantics as discussed above to allow for locking of a function provided as a generator

CI is green.

File: cmd/nerdctl/volume_remove_linux_test.go
Comment (AkihiroSuda) [reply cmd/nerdctl/volume_remove_linux_test.go:44]: I'm not sure about this. Easier to merge this PR if this part can be omitted.

Comment (apostasie) [reply cmd/nerdctl/volume_remove_linux_test.go]: Ok. Omitted with latest push.

**Comment by apostasie** [comment]
Rebased. Pending green CI.

File: pkg/store/filestore.go
Comment (AkihiroSuda) [reply pkg/store/filestore.go]: This should be removed now?

**Comment by AkihiroSuda** [comment]
`GroupRename()` seems added (and commented out) in the first commit and then removed in the third commit.
Probably this should be just removed from the first commit

**Comment by apostasie** [comment]
> `GroupRename()` seems added (and commented out) in the first commit and then removed in the third commit. Probably this should be just removed from the first commit

My bad. Thanks for catching it.
Last push corrects it.

(CI failing on Hub being 500)

File: pkg/store/store.go
Comment (fahedouch) [inline pkg/store/store.go:93]:
```suggestion
// Locker describes a locking mechanism that can be used to encapsulate operations in a safe way
```

**Comment by fahedouch** [comment]
big UP for this PR, I have to points to raise here:
- please avoid huge PR
-  we may need a doc for this abstraction (e.g big picture system design)

**Comment by apostasie** [comment]
> big UP for this PR 👍 , I have two points to raise here:
> 
> * please avoid huge PR, it is really hard to review
> * we might need a document for this abstraction (e.g. big picture abstraction design)

Thanks a lot @fahedouch 
Really appreciate the sharp eye review
On my phone now but will address your comments momentarily

ack on splitting in more commits
Will definitely do in upcoming work

File: pkg/cmd/volume/prune.go
Comment (apostasie) [reply pkg/cmd/volume/prune.go:48]: Mmmm... I see.
Well, we do still need the `usedVolumes` list to display a warning to the user if they ask to rm an in-use volume.

Your point still stand though, we might want to cleanup things a bit more under `pkg/cmd` later on.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Review by fahedouch — APPROVED** [review]
Thanks

**Pull request merged**
